# Assignment 5

### Author : Omer Ozeren

### Introduction

1. Choose a corpus of interest.
2. How many total unique words are in the corpus? (Please feel free to define unique words in any interesting,
defensible way).
3. Taking the most common words, how many unique words represent half of the total words in the corpus?
4. Identify the 200 highest frequency words in this corpus.
5. Create a graph that shows the relative frequency of these 200 words.
6. Does the observed relative frequency of these words follow Zipf’s law? Explain.
7. In what ways do you think the frequency of the words in this corpus differ from “all words in all corpora.”

In [2]:
import nltk, re, pprint
from nltk import word_tokenize
from nltk.corpus import stopwords
from nltk.corpus import webtext
import urllib.request as ur
import pandas as pd
import matplotlib.pyplot as plot
import string as str
from nltk.stem import PorterStemmer
%matplotlib inline

### 1. Choose a Corpus of Interest

I decided to use Project Guttenberg that it is a site which allows you to read books directly from the HTML.I choose the most popular data from website.The data can be found [here](http://www.gutenberg.org/files/84/84-h/84-h.htm). 

In [4]:
url="http://www.gutenberg.org/files/84/84-h/84-h.htm"
path = ur.urlopen(url)
raw = path.read().decode('utf8')

### 2. How many total unique words are in the corpus? (Please feel free to define unique words in any interesting, defensible way).

In [5]:
starting_position = raw.find("></a>Chapter 1</h2>\r\n\r\n<p>\r\n")
ending_position = raw.find("*** END OF THE PROJECT GUTENBERG EBOOK FRANKENSTEIN ***")
raw = raw[starting_position:ending_position]

In [7]:
raw_token = word_tokenize(raw)

len(set(raw_token))

7298

We have 7,298 defined words in the raw text file data. I will only take alpha numeric words and make all words lower case.

In [12]:
unique_words = [word for word in raw_token if word.isalpha()] #alphanumeric

unique_words = [word.lower() for word in unique_words] #lowercase

In [13]:
len(set(unique_words))

6613

In [16]:
unique_words[:10]

['chapter', 'p', 'i', 'am', 'by', 'birth', 'a', 'genevese', 'and', 'my']

We have 6,613 unique words.However, The unique words include letters which needs to be excluded from data.I will calculate the lenght of each item in unique words.

In [24]:
df_words_len = pd.DataFrame({'words':unique_words}) # from list to DataFrame
df_words_len['length'] = df_words_len['words'].apply(lambda x: len(x))
df_words_len

words  length
0       chapter       7
1             p       1
2             i       1
3            am       2
4            by       2
...         ...     ...
71221  distance       8
71222       end       3
71223   chapter       7
71224       div       3
71225     block       5

[71226 rows x 2 columns]

By looking at the DataFrame above, I can see that i need to exclude the words that have lenght 1.

In [36]:
exclude_words_df= (df_words_len[df_words_len['length']==1].sort_values(by='length'))
exclude_words_df.head()

words  length
1         p       1
49085     i       1
49068     a       1
49057     p       1
49039     i       1

In [43]:
df_words_clean = df_words_len[df_words_len.length > 1]
df_words_clean

words  length
0       chapter       7
3            am       2
4            by       2
5         birth       5
7      genevese       8
...         ...     ...
71221  distance       8
71222       end       3
71223   chapter       7
71224       div       3
71225     block       5

[66459 rows x 2 columns]

In [42]:
unique_words=list(df_words_clean['words'])
len(unique_words)

71226

After removing words that has 1 characters, I can see that the total unique words in the corpus is **66,459.**

### 3.Taking the most common words, how many unique words represent half of the total words in the corpus?

In [ ]:
data_freq_dist = nltk.FreqDist(unique_words)
data_freq_dist